### DISCUSSION TITLE
[FE] babel 도입 고려

### DISCUSSION THREAD

**Discussion by Jaymyong66** [woowacourse-teams/2024-code-zap#431] [이야기해봐요]
babel을 도입해보고 싶은 니즈가 생겨서 발제해봅니다.
주의) 꼭 하고 싶다 - X, 생각보다 overhead가 클 수 있음

babel이란? [테코톡 영상](https://www.youtube.com/watch?v=o-5K5Sc7L1k)

---

- 장점 1.
현재 쓰는 data-testis를 production에서 지울 수 있다.
이게 babel의 플러그인임. 
https://github.com/coderas/babel-plugin-jsx-remove-data-test-id
[관련 아티클 추가](https://dpericich.medium.com/removing-data-test-attributes-from-react-production-dom-5ea4ea018acc)

- 장점 2. 
emotion의 해시 처리되는 className을 어느 정도 보이게 할 수 있다.
공홈 플러그인의 autoLabel 옵션
https://emotion.sh/docs/@emotion/babel-plugin
<img width="300" alt="스크린샷 2024-08-13 오후 7 16 48" src="https://github.com/user-attachments/assets/edf36d97-e82d-41cd-a1a0-f0c754915d40">

해당 장점은 개발할 때도 유용하지만 Sentry에서 더 큰 유용성을 발견했다.
<img width="500" alt="스크린샷 2024-08-13 오후 7 17 30" src="https://github.com/user-attachments/assets/bfb0a913-a9b6-41fa-bd29-b93dbfec8812">

---

- 단점 1
ts-loader를 babel-loader로 바꾸고 필요하다면 type check하는 플러그인을 써야한다(babel-loader는 기본적으로 타입 에러는 봐주지 않기에)

- 단점 2
지금 설치되어있는 라이브러리들과 환경 간의 의존성 관련 어떤 문제가 발생할지는 미지수.

- 단점 3
위 장점 1, 2가 사실은 DX에만 관련이 있고 사용자에게는 문제가 없다.
할 일이 많은데 babel 도입하면서 트러블 슈팅할 시간을 할애해야한다.
(예를 들어 emotion에는 수동으로 중요한 styled component에만 label을 달아주면 된다.)
(data 어트리뷰트를 test용으로 쓰는 것은 보안이나 뭐나 전혀 문제가 없다)

**Comment by Hain-tain** [comment]
- 옮겨보는 것은 좋습니다! 특히 장점2 가 UI 확인하기 좋아보이네요! 다만, 지금 당장 급한 이슈는 아니기 때문에 Z 기간동안은 지금 상황을 지켜보고 A 즈음에 시도해보아도 좋을 것 같습니다! 아니면 아예 레벨 4 들어가면서 해봐도 좋을 것 같아요!
- 개인적으로 dev 서버만 바벨로 옮기면 어떨까 싶어요..! (이게 더 시간이 많이 들지도 모르지만..ㅎㅎ) 그러면 타입 체크는 지금처럼 ts-loader로 할 수 있고, production으로 배포되는 서버는 타입 에러 이미 dev에서 잡았을 거라고 생각하고 그냥 넘어가는거죠..! 그러면 빌드 시간도 조금 더 빨라지지 않을까 생각합니다.(제가 알기론 ts-loader 보다 babel이 더 빠르다고 하더라고요!)
